lib/errorUtils.js

```diff
@@ -0,0 +1,30 @@
+'use strict'
+
+const Assert = require('assert')
+
+const systemErrors = [
+
+  // JavaScript
+  EvalError,
+  RangeError,
+  ReferenceError,
+  SyntaxError,
+  TypeError,
+  URIError,
+
+  // Node
+  Assert.AssertionError
+]
+
+function isSystem (err) {
+  for (var i = 0; i < systemErrors.length; i++) {
+    var type = systemErrors[i]
+    if (err instanceof type) {
+      return true
+    }
+  }
```

Comment:
This can be avoided or at least delayed: `err.constructor.name` can be used to check if in a map the name exists. I don't understand why we need to check the inheritance instead of the constructor
```
const systemErrorMap = {
  EvalError: true,
  RangeError: true,
  ReferenceError: true,
  SyntaxError: true,
  TypeError: true,
  URIError: true,

  Assert.AssertionError: true
}

function isSystem (err) {
  return err && systemErrorMap[err.constructor.name]
}
```
Anyway, I don't know if AssertionError should be listed here.

Reply:
`instanceof` check is more reliable and you can derive from these types.

```js
class A extends TypeError {}
const a = new A()
const t = new TypeError()
console.log(a.constructor.name) // "A"
console.log(t.constructor.name) // "TypeError"

console.log(a instanceof Error) // true
console.log(t instanceof Error) // true
```

> Anyway, I don't know if AssertionError should be listed here.

I would say because of a `AssetionError` signal a wrong use and not a business error.

Reply:
In your example, I don't know why `a` error shouldn't be considered as a business error. 

If a developer writes an error class that inherits from a "System" error I consider it as a business error.

Reply:
As documented in the "error-handling" section it's best practice to build your own error hierarchy instead to express your business errors with these types. If you use an `AssertionError` it typically means that you use the interface in a wrong way. You can still throw any ``Error` you want.

> If a developer writes an error class that inherits from a "System" error I consider it as a business error.

No, it has just a different semantic but is classified into a system error.

```js
class OrderTypeError extends TypeError {}
function processOrder(order) {
  if(typeof order !== 'object') throw new OrderTypeError('order should be an object')
}
```

Reply:
From my point of view that if can be seen as a business logic check.

nodejs is not java, `OrderTypeError` can be a standard error with a `code` property. Combined with `setErrorHandler` you can achieve a nice HTTP output.

Anyway the `instanceof` check can be made after the map check

Reply:
Why do you want to limit that? It sounds much cleaner to me to say that any derivation from a system error is a system error. If you want to build a business error just use `Error`. Let us not drown in super micro-optimizations for an error case.

Reply:
Because the `A` error isn't a "system" error. Bad access, invoke a non function etc... are. From my point of view every other errors are business logic ones. Only "runtime" errors or nodejs core errors aren't.

Reply:
`A` should just show that it is possible to derivate from it.

> From my point of view every other errors are business logic ones. Only "runtime" errors or nodejs core errors aren't.

The list contains exactly these errors https://github.com/nodejs/node/blob/a82b1b7b30d52cd6a3ea9f129cdc3d83e0370a34/lib/internal/errors.js#L231

Reply:
I also understand your point but it's a new way to handle with errors in promises and async / await.

Reply:
Have you read ?

- https://github.com/fastify/fastify/pull/591#pullrequestreview-85897548
- https://github.com/fastify/fastify/pull/591#issuecomment-354457343